runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp

```diff
@@ -78,6 +79,82 @@ void J9::RecognizedCallTransformer::process_java_lang_Class_IsAssignableFrom(TR:
    fromClass->recursivelyDecReferenceCount();
    }
 
+static void substituteNode(
+   TR::NodeChecklist &visited, TR::Node *subOld, TR::Node *subNew, TR::Node *node)
+   {
+   if (visited.contains(node))
+      return;
+
+   visited.add(node);
+
+   TR_ASSERT_FATAL(node != subOld, "unexpected occurrence of old node");
+
+   for (int32_t i = 0; i < node->getNumChildren(); i++)
+      {
+      TR::Node *child = node->getChild(i);
+      if (child == subOld)
+         {
+         subOld->recursivelyDecReferenceCount();
```

Comment:
I have wondered it is a safer idiom to `incReferenceCount/setAndIncChild` first and then call `recursivelyDecReferenceCount` later to account for the case when there may be nodes that are shared in the subtree of the child that is being diconnected/connected (to avoid some node in the subtree being left with ref count 0). Maybe this is not possible in this specific case, but I thought I'd mention it anyway.

Reply:
I think that's a good suggestion in general. This particular case is a bit weird because `subOld` won't have its refcount reach zero here. The methods that decrement the refcount both specify what happens at zero (though for one of them, it's just nothing), so we don't really have a nice way to decref without choosing one of those two behaviours. To make this clearer, I could put in an assertion and a comment, e.g.
```
// subOld occurs at least at its point of evaluation and here
TR_ASSERT_FATAL(subOld->getReferenceCount() >= 2, "...");
subOld->decReferenceCount();
```

Reply:
I've gone ahead and made that change

Reply:
Thanks for adding the assert, I agree that it is okay if the `subOld` reference count cannot reach zero here.